core/src/main/scala/neotypes/Session.scala

```diff
@@ -1,40 +1,45 @@
 package neotypes
 
-import neotypes.implicits.syntax.async._
-import neotypes.utils.stage._
+import internal.syntax.async._
+import internal.syntax.stage._
+
 import org.neo4j.driver.v1.{Session => NSession, Transaction => NTransaction}
 
 import scala.language.higherKinds
 
-final class Session[F[_]](session: NSession)(implicit F: Async[F]) {
-  def beginTransaction(): F[Transaction[F]] =
-    F.async { cb =>
-      session
-        .beginTransactionAsync()
-        .accept { tx: NTransaction => cb(Right(new Transaction(tx)(F))) }
-        .recover { ex: Throwable => cb(Left(ex)) }
-    }
+final class Session[F[_]](private val session: NSession) extends AnyVal {
+  def transaction[R[_]](implicit F: Async.Aux[F, R]): R[Transaction[F]] =
+    F.resource(createTransaction) { transaction => transaction.close }
 
-  def close(): F[Unit] =
+  private[neotypes] def createTransaction(implicit F: Async[F]): F[Transaction[F]] =
     F.async { cb =>
       session
-        .closeAsync()
-        .accept { _: Void => cb(Right(())) }
+        .beginTransactionAsync()
```

Comment:
I thought methods with empty parens were gone?

Reply:
Those are from the **Java** API, as such, those are real side-effects!

Reply:
Fair enough! :smile: